src/core/exec-invoke.c

```diff
@@ -5267,19 +5277,29 @@ int exec_invoke(
         log_command_line(context, params, "Executing", executable, final_argv);
 
         if (params->exec_fd >= 0) {
-                usec_t t = now(CLOCK_MONOTONIC);
+                uint8_t x = 1;
 
                 /* We have finished with all our initializations. Let's now let the manager know that. From this point
-                 * on, if the manager sees POLLHUP on the exec_fd, then execve() was successful. We send a
-                 * timestamp so that the service manager and users can track the precise moment we handed
-                 * over execution of the service to the kernel. */
+                 * on, if the manager sees POLLHUP on the exec_fd, then execve() was successful. */
 
-                if (write(params->exec_fd, &t, sizeof(t)) < 0) {
+                if (write(params->exec_fd, &x, sizeof(x)) < 0) {
                         *exit_status = EXIT_EXEC;
                         return log_exec_error_errno(context, params, errno, "Failed to enable exec_fd: %m");
                 }
         }
 
+        if (params->handoff_timestamp_fd >= 0) {
+                /* Best effort: get a timestamp now and send it upstream */
+                struct dual_timestamp dt;
+                dual_timestamp_now(&dt);
+
+                (void) send(params->handoff_timestamp_fd,
```

Comment:
The reason I used the existing pipe is that it makes it trivial to ensure it is resilient - either the activation notification _and_ the timestamp are received and recorded, or neither is. But with this implementation that just ignores errors when sending the timestamp the metric is no longer reliable, and there will be cases where the payload is started with no error, but there's no metric - or worse, the old metric is published, I haven't seen where it would be reset if it's there. Sorry, but this is a regression from the current implementation, to be useful it needs to be reliable.

So, if you really want to make it so much more complex with a new protocol and all that, that is fine by me, but it needs to be as resilient and atomic as the current implementation. Hence, please change it to check for the error rather than ignore it, and if it goes wrong the unit needs to fail, and the exec_fd needs to be disarmed too. And same on the receiving end, either both protocols work or none do, if they are enabled.

Reply:
I think it's reasonable to fail if the send fails. It shouldn't fail, and it'll be easier to figure out any bugs if we make it fail hard.

Reply:
Come on, it's not going to be atomic *ever*. execve() doesn't allow that.

We first have to take the timestamp, and then send it, and *then* we can call execve(). No matter what. Also, the atomicity here is irrelevant anyway the clocks we are querying aren't necessarily quick enough anyway.

And there is no "regression" here anyway, we never released the code.

Also, an old metric cannot be published, because we reset all ExecStatus objects in service_start(), we always start fresh.

Reply:
> Come on, it's not going to be atomic _ever_. execve() doesn't allow that.

Let me explain again what I mean by that: of course it's not in the sense of being atomic with the syscall, as that's obviously impossible. I meant in the sense of state tracking vs metric recording, which are purely internal.

The version currently in main either tracks the unit as successful _and_ records the metric, _or_ tracks the unit as failed to exec and invalidates the metric. The version in this PR loses this guarantee, and that's a regression w.r.t. the current state in main, and it is a functional change that negatively affects real-world usage, because suddenly you have to take into account that you might have services running that should have a valid metric, but don't.

It's understandable why: it's much harder to do so with two independent pipes, tracked and consumed and produced independently, which, again, it's the reason why I simply used the same one. But it's not impossible to make it safe with multiple pipes, and given you really want to do the latter, that is fine by me, but again, please ensure the same property is provided in the new implementation, as it was in the current one. If the second pipe fails, fail the unit and rollback the exec_fd message, and viceversa on the other side.

> Also, an old metric cannot be published, because we reset all ExecStatus objects in service_start(), we always start fresh.

That's perfect, thanks for clarifying.